### ISSUE TITLE
Support multiple GitHub accounts

### ISSUE THREAD

**Issue by marvint24** [microsoft/vscode-copilot-release#109]
I use multiple GitHub accounts. One has access to GitHub Copilot and GitHub Copilot Chat, and the other one does not.

After installing the GitHub Copilot and GitHub Copilot Labs extensions, you are asked to sign in with your account. 
After installing the GitHub Copilot Chat extension, you are not.

There is also no option to manage it under:
1.  ![image](https://github.com/microsoft/vscode-copilot-release/assets/123128755/c631c74c-6eca-41cc-8f1e-81f039dbcea2)
2.  ![image](https://github.com/microsoft/vscode-copilot-release/assets/123128755/eca0f3df-61c1-4df8-af92-c3d360c47bdf)
3.  ![image](https://github.com/microsoft/vscode-copilot-release/assets/123128755/0e91b77b-e526-4008-9624-7f96e5e44f79)

At the moment, I only get this error message:
![image](https://github.com/microsoft/vscode-copilot-release/assets/123128755/2e2be985-c7c9-4724-abbc-a6651a7b6196)

Please add an option to select the account.

**Comment by TylerLeonhardt** [comment]
Depends on https://github.com/microsoft/vscode/issues/127967

**Comment by Falven** [comment]
This is a critical enterprise scenario... Data isolation of enterprise account, potentially even customer data. We have been waiting on this for months. Please prioritize this already.
It could easily be managed by the VS Code Trusted Extensions functionality if Copilot respected those settings.

**Comment by LogicDaemon** [comment]
> Depends on [microsoft/vscode#127967](https://github.com/microsoft/vscode/issues/127967) and then the copilot extension will need to take advantage of this.

but the main Copilot extension already works fine with a separate GitHub account! And even Copilot Labs worked.

But the Copilot Chat extension ignores the account selected for Copilot and just goes for the main account used in VS Code for settings sync!

**Comment by LogicDaemon** [comment]
Recently, the main Copilot extension started complaining about the github account and stopped working.

I ended up setting a separate Microsoft account just for VS Code settings sync.
This way, I can authenticate in Github with Copilot account at work, and keep the settings in a separate Microsoft account available on all my computers.

**Comment by Twoleeland** [comment]
> > Depends on [microsoft/vscode#127967](https://github.com/microsoft/vscode/issues/127967) and then the copilot extension will need to take advantage of this.
> 
> but the main Copilot extension already works fine with a separate GitHub account! And even Copilot Labs worked.
> 
> Copilot Chat extension just ignores the account selected for Copilot and instead goes for the main account used in VS Code for settings sync!
> 
> I don't want to switch between VS Code profiles. I like having same settings at work and at home.
> 
> But I only have Copilot enabled for the work account, which I don't use at home.

For now, it's a possible workaround for some developers, but not a “real” solution. About 1.5 years ago, syncing settings with Microsoft account was suddenly a problem. After days and weeks of people here on GitHub complaining about sync issues with MS accounts and MS not doing anything about it, I switched to GitHub sync. Since then, there have been no more problems with synchronizing settings.

Anyway, they should split Copilot-licensing from Copilot-settings or introduce a possibility to link private accounts to business accounts regarding the Copilot-licensing in some way.

**Comment by ian-h-chamberlain** [comment]
Now that https://github.com/microsoft/vscode/pull/222131 has merged in VSCode, is there a plan to implement support for it in this extension? I still have not found a reliable way to use a different account for GitHub Pull Requests vs this extension, particularly after some of the changes around #1444 (I guess `github.copilot.preferredAccount` is not supposed to be set at all, confusingly?)

**Comment by TylerLeonhardt** [comment]
@ian-h-chamberlain have you tried:
```
"github.experimental.multipleAccounts": true
```
* Signing in to GH PR with account 1
* Setting:
```
"github.copilot.preferredAccount":"YOUR ACCOUNT 2 USERNAME"
```
* Signing in to GitHub Copilot Chat with account 2?

**Comment by ian-h-chamberlain** [comment]
@TylerLeonhardt I just tried these steps; I guess maybe my order of operations was part of the problem before?

1. Configure `github.experimental.multipleAccounts` to true
1. Signed out of all accounts
1. Deleted `preferredAccount` from configuration
1. Sign into to GH PR extension
1. Configured `preferredAccount` with my company-provisioned username `$companyUser`
1. Sign into GH Copilot + Copilot Chat

Now, Copilot Chat works as expected but not Copilot itself, with several of these logs in its output channel:
```
2024-09-09 13:06:07.000 [info] [auth] Invalid copilot token: missing token: 403 
2024-09-09 13:06:07.026 [info] [auth] Invalid copilot token: missing token: 403 
```

I can see that Chat is using the correct account name, but Copilot is not, based on the timestamps in the "trusted extension" view:

### GH PR account (`ian-h-chamberlain (GitHub)`)
![Screenshot 2024-09-09 at 13 09 49](https://github.com/user-attachments/assets/678c6fd6-dbd6-462a-ab94-4897a891cf77)

### `preferredAccount` with Copilot license (`$companyUser (GitHub)`)
![Screenshot 2024-09-09 at 13 09 42](https://github.com/user-attachments/assets/34497bab-823f-44de-9ebb-e0277113ea6e)

There also appears to be a third account `$companyUser (GitHub Copilot Chat)` which has no extensions associated with it:
![Screenshot 2024-09-09 at 13 15 25](https://github.com/user-attachments/assets/237c6615-4dde-4edc-9ebc-a8e8adbedadb)

**Comment by TylerLeonhardt** [comment]
Couple things here that would fix the issue for you:
* Copilot (not Chat) doesn't honor the `preferredAccount` setting yet... it should and we're going to adopt it there... working with the right folks to do that work
* This issue https://github.com/microsoft/vscode/issues/225943 should also probably fix the issue for you, which I'm working on this month

**Comment by ian-h-chamberlain** [comment]
> 3. Sign in to GitHub Copilot with the correct account
> 4. Sign in to GitHub Pull Request with its correct account

I think this is the part where it becomes problematic... after sign-in it presents this prompt:

![Screenshot 2024-09-09 at 14 07 19](https://github.com/user-attachments/assets/6839b837-02dd-409f-b559-04966bcf4366)

1. Click `Allow`:

   Once the second sign-in from the above prompt completes, there seems to be no way to separately sign into GitHub Pull Request, because they are treated under the same "GitHub" umbrella? 

   `GitHub Pull Requests: Sign in to GitHub` has no effect at this point, which I guess is maybe due to GHPR not supporting multiple accounts either (https://github.com/microsoft/vscode-pull-request-github/issues/5159, https://github.com/microsoft/vscode-pull-request-github/issues/3595)

2. Click `Cancel`:

   Both Copilot and Chat seem to work correctly at first. Proceeding to steps 4-6 sort of seems to work at this point? Although I didn't need to re-sign in at step 6 because chat already appeared to be enabled correctly.

   However, Copilot completions working seems to be temporary; I'm presented with this immediately
![Screenshot 2024-09-09 at 14 09 53](https://github.com/user-attachments/assets/3e9ff102-10a9-49f2-85f4-dbc57ce4ea3e) 
   although it seems to be kind of a mistaken warning at first. After reloading the VSCode window, the warning appears again and completions stops working.

   Chat authentication seems to persist across a reload.

----

I'm glad to hear there is planned work being done here! I think Copilot (not Chat) using `preferredAccount` would likely solve this, although it might be worth posing a similar idea to the GH Pull Requests folks for that extension just to make everything explicit. Maybe some kind of top-level `github.preferredAccount` as a fallback would make sense too, as part of the `experimental.multipleAccounts` feature?

**Comment by TylerLeonhardt** [comment]
Update in Core:
https://github.com/microsoft/vscode/issues/127967#issuecomment-2364272689

Actually I think with that change, this should be supported.

**Comment by claidler** [comment]
> Update in Core: [microsoft/vscode#127967 (comment)](https://github.com/microsoft/vscode/issues/127967#issuecomment-2364272689)
> 
> Actually I think with that change, this should be supported.

This resolves the issue for me!  Thank you

**Comment by TylerLeonhardt** [comment]
@claidler you shouldn't have to do that and I don't repro that behavior. Can you open a new issue please with some solid steps that you follow?

**Comment by TylerLeonhardt** [comment]
Closing this issue since the feature is resolved.
